File: app/App_Resources/Android/src/main/java/com/helpers/Metronome.java
package com.helpers;

import com.helpers.Beep;
import java.util.Timer;
import java.util.TimerTask;
import android.util.Log;

public class Metronome {
  int speed;
  Beep mainTone;
  Beep secondaryTone;
  Timer timer;
  TimerTask tick;
  boolean isRunning = false;

  public Metronome(int bpm, Beep mainBeep, Beep secondaryBeep) {
    this.speed = bpm;
    this.mainTone = mainBeep;
    this.secondaryTone = secondaryBeep;
    this.timer = new Timer();
    this.tick = new TimerTask() {
      @Override
      public void run() {
        mainTone.play();
      }
    };
  }

  public void changeSpeed(int bpm) {
    this.speed = bpm;

    if (this.isRunning) {
      this.stop();
      this.timer.purge();
      this.start();
    }
  }

  public void start() {
    this.timer.scheduleAtFixedRate(this.tick, 0, 60000 / this.speed);
    this.isRunning = true;
  }

  public void stop() {
    this.timer.cancel();
    this.isRunning = false;
  }
}

File: app/lib/Metronome.js
import {setInterval, clearInterval} from '@nativescript/core/timer';

export default class Metronome {
  constructor(bpm, timeSignature = 4) {
    this.speed = bpm;
    this.timeSignature = timeSignature;

    this.mainTone = new com.helpers.Beep(1000, 60);
    this.secondaryTone = new com.helpers.Beep(800, 60);
    this.isRunning = false;
    this.currentTick = 0;
  }

  changeSpeed(bpm) {
    this.speed = bpm;

    if (this.isRunning) {
      this.stop();
      this.start();
    }
  }

  start() {
    this.timer = setInterval(() => {
      this.nextTone();
    }, 60000 / this.speed);

    this.isRunning = true;
  }

  stop() {
    clearInterval(this.timer);

    this.isRunning = false;
    this.currentTick = 0;
  }

  nextTone() {
    const beat = this.currentTick % this.timeSignature;

    if (beat === 0) {
      this.mainTone.play();
    } else {
      this.secondaryTone.play();
    }

    this.currentTick++;
  }
}
